source/common/upstream/load_balancer_impl.cc

```diff
@@ -890,6 +892,140 @@ HostConstSharedPtr RandomLoadBalancer::peekOrChoose(LoadBalancerContext* context
+      shuffle_sharder_(ShuffleSharder<HostConstSharedPtr>(12345)) {
+
+  // ENVOY_LOG(debug, "endpoints_per_cell: {}", endpoints_per_cell_);
+  // ENVOY_LOG(debug, "use_zone_as_dimension: {}", use_zone_as_dimension_);
+
+  for (auto dimension : config.dimensions())
+    dimensions_.push_back(dimension);
+  if (use_zone_as_dimension_)
+    dimensions_.push_back("_envoy_zone");
+  if (!dimensions_.size())
+    dimensions_.push_back("_envoy_unit_coord");
+
+  lattice_ = new Lattice<HostConstSharedPtr>(dimensions_);
+
+  priority_update_cb_ =
+      priority_set_.addPriorityUpdateCb([this](uint32_t /*priority*/, const HostVector& hosts_added,
+                                               const HostVector& hosts_removed) -> void {
+        add_hosts(hosts_added);
+        remove_hosts(hosts_removed);
+      });
+}
+
+HostConstSharedPtr ShuffleShardLoadBalancer::chooseHostOnce(LoadBalancerContext* context) {
+  absl::optional<uint64_t> hash;
+  if (context) {
+    hash = context->computeHashKey();
+  }
+  const uint64_t seed = hash ? hash.value() : random_.random();
+
+  auto endpoints =
+      shuffle_sharder_.shuffleShard(*lattice_, seed, endpoints_per_cell_)->get_endpoints();
+
+  // Random
+  if (lb_type_ == LoadBalancerType::Random)
+    return endpoints[random_.random() % endpoints.size()];
+
+  // Least Request
+  HostConstSharedPtr candidate_host = nullptr;
+  for (uint32_t choice_idx = 0; choice_idx < 2; ++choice_idx) {
+    const int rand_idx = random_.random() % endpoints.size();
+    HostConstSharedPtr sampled_host = endpoints[rand_idx];
+
+    if (candidate_host == nullptr) {
+      // Make a first choice to start the comparisons.
+      candidate_host = sampled_host;
+      continue;
+    }
+
+    const auto candidate_active_rq = candidate_host->stats().rq_active_.value();
+    const auto sampled_active_rq = sampled_host->stats().rq_active_.value();
+    if (sampled_active_rq < candidate_active_rq) {
+      candidate_host = sampled_host;
+    }
+  }
+
+  return candidate_host;
+}
+
+void ShuffleShardLoadBalancer::remove_hosts(const HostVector& hosts) {
+  for (auto& host : hosts) {
+    auto coord = get_coord(host);
+    if (!coord) {
+      continue;
+    }
+
+    std::vector<HostConstSharedPtr> hosts;
+    hosts.push_back(host);
+    lattice_->remove_endpoints_for_sector(*coord, hosts);
+  }
+}
+
+void ShuffleShardLoadBalancer::add_hosts(const HostVector& hosts) {
+  for (auto& host : hosts) {
+    auto coord = get_coord(host);
+    if (!coord) {
+      continue;
+    }
+
+    std::vector<HostConstSharedPtr> hosts;
+    hosts.push_back(host);
+    lattice_->add_endpoints_for_sector(*coord, hosts);
+  }
+}
+
+absl::optional<std::vector<std::string>>
+ShuffleShardLoadBalancer::get_coord(const HostConstSharedPtr& host) {
+  std::vector<std::string> coord;
+
+  if (use_dimensions_) {
+    if (!host->metadata()) {
+      // ENVOY_LOG(warn, "ignoring host {} because it has no metadata", host->hostname());
+      return {};
```

Comment:
does this return an empty optional or an optional with an empty list?

Reply:
My understanding is that this returns an empty optional. It looks like `std::nullopt` is used in other portions of envoy and may be more idiomatic. Will switch.